File: lib/arithmetic.cpp
#include "number.hpp"


Number Number::operator+(Number b) const {
  Number res;
  res.digits.clear();

  int carry = 0;
  size_t maxsize = std::max(digits.size(), b.digits.size());

  for (size_t i = 0; i < maxsize || carry; ++i) {
    int sum = carry;
    if (i < digits.size()) sum += digits[i];
    if (i < b.digits.size()) sum += b.digits[i];
    res.digits.push_back(sum % 10);
    carry = sum / 10;
  }
  return res;
}

Number Number::operator*(Number b) const {
  Number res = *this;
  for(Number counter(1); counter != b; counter = counter + Number(1))
    res = res + *this;
  return res;
}

#include "arithmetic.hpp"

Number pow(Number a, Number b){
  Number res = a;
  for(Number k(1); k != b; k = k + Number(1)){
    res = res*a;
  }
  return res;
}

File: lib/arithmetic.mjs
// Internal: convert character to corresponding digit value
function covToDig(n, i){
  return n[i].charCodeAt(0) - '0'.charCodeAt(0);
}

// returns n + b
export const add = (n, b) => {
  if(typeof n != 'string' && typeof b != 'string'){
    n = n.toString();
    b = b.toString();
  }
  let result = ""; let carry = 0;

  // Pad the shorter string with leading zeros
  while (n.length < b.length) n = "0" + n;
  while (b.length < n.length) b = "0" + b;

  // Add digits from right to left
  for (let i = n.length - 1; i >= 0; i--) {
    let sum = covToDig(n,i) + covToDig(b,i) + carry;

    // Compute the new digit and carry
    let nd = sum % 10;
    carry = (sum-(nd))/10;  // Manual floor operation for carry

    // Convert the new digit back to a character and prepend it to the result
    result = String.fromCharCode(nd+'0'.charCodeAt(0)) + result;
  }

  // If there's a carry left, add it to the result
  if (carry > 0) {
    result = String.fromCharCode(carry + '0'.charCodeAt(0)) + result;
  }

  return result;
}

// unfortunately this is also required by tetration
export const sub = (n,b) => {
  // STUPID!
  if(typeof n != 'string' && typeof b != 'string'){
    n = n.toString();
    b = b.toString();
  }


  let res = "";
  let borrow = 0;

  // Pad the shorter string with leading zeros
  while (b.length < n.length) b = "0" + b;

  // Subtract digits from right to left
  for (let i = n.length - 1; i >= 0; i--) {
    let diff = covToDig(n,i) - covToDig(b,i) - borrow;

    if (diff < 0) {
      diff += 10; borrow = 1;
    } else borrow = 0;

    res = String.fromCharCode(diff + '0'.charCodeAt(0)) + res;
  }

  // Remove leading zeros
  while (res.length > 1 && res[0] === "0") {
      res = res.substring(1);
  }

  return res;
}

// returns n*b
export const mult = (n, b) => {
  let res = "0";
  for(let counter = "0"; counter != b; counter = add(counter,"1")) {
    res = add(res, n.toString());
  }
  return res;
}

export const exp = (n,b) => {
  let res = "1";
  for(let counter = "0"; counter != b; counter = add(counter, "1")){
    res = mult(res, n.toString());
  }
  return res;
}